docs: update plugin documentation (#888)
jackyzha0/quartz: docs/configuration.md

![[quartz transform pipeline.png]]

```ts title="quartz.config.ts"
plugins: {
  transformers: [...],
  filters: [...],
  emitters: [...],
}
```

- [[tags/plugin/transformer|Transformers]] **map** over content (e.g. parsing frontmatter, generating a description)
- [[tags/plugin/filter|Filters]] **filter** content (e.g. filtering out drafts)
- [[tags/plugin/emitter|Emitters]] **reduce** over content (e.g. creating an RSS feed or pages that list all files with a specific tag)

> You can customise the behaviour of Quartz by adding, removing and reordering plugins in the `transformers`, `filters` and `emitters` fields.

Reviewer: Oh I also forgot to mention.

Probably worth saying here is that the order of items in the array matters, atm.

So i.e: FrontMatter should always get set first, Latex second, etc.

@jackyzha0

Author: It's right next to this sentence:

![image](https://github.com/jackyzha0/quartz/assets/1475672/588a58dd-bbe9-46d9-9c0c-b7b05b3c911e)

Reviewer: oh i didn't catch that. Can you highlight `position sensitive`, or maybe I guess the term is `location-dependent` 😄

> [!note]
> Each node is modified by every transformer _in order_. Some transformers are position sensitive, so you may need to pay particular attention to whether they need to come before or after certain other plugins.

> To add a plugin, you insert a line with its function call, followed by a semicolon in the appropriate category. For example, to add the [[ExplicitPublish]] plugin (a [[tags/plugin/transformer|Transformer]], you would add the following line:

Reviewer 2: ```suggestion
You should take care to add the plugin to the right entry corresponding to its plugin type. For example, to add the [[ExplicitPublish]] plugin (a [[tags/plugin/transformer|Transformer]], you would add the following line:
```

```ts title="quartz.config.ts"
transformers: [
  ...
  Plugin.ExplicitPublish(),
  ...
],
```

> To remove a plugin, you remove all occurrences of it in the `quartz.config.ts`.

Reviewer 2: ```suggestion
To remove a plugin, you should remove all occurrences of it in the `quartz.config.ts`.
```

> In addition, plugins may also have their own configuration settings that you can pass in. If you do not pass in a configuration, the plugin will use its default settings.

Reviewer 2: ```suggestion
To customize plugins further, some plugins may also have their own configuration settings that you can pass in. If you do not pass in a configuration, the plugin will use its default settings.
```

]
```

The current default plugins can be found [here](https://github.com/jackyzha0/quartz/blob/v4/quartz.config.ts). For a list of all plugins and the available configuration options, visit the [[plugin]] tag overview.

> If you'd like to make your own plugins, read the guide on [[making plugins]] for more information.

Reviewer: ```suggestion
If you'd like to make your own plugins, read the guide on [[making plugins|making custom plugins]] for more information.
```

Author: Done.

```

If you'd like to make your own plugins, read the guide on [[making plugins]] for more information.

> By default, Quartz is configured to use the following plugins:

Reviewer: ```suggestion
By default, default plugins can be found [here](https://github.com/jackyzha0/quartz/blob/v4/quartz.config.ts)
```

Future proof ourself

Author: Done.

> ```ts
> transformers: [
>   Plugin.FrontMatter(...),
>   Plugin.CreatedModifiedDate(...),
>   Plugin.Latex(...),
>   Plugin.SyntaxHighlighting(...),
>   Plugin.ObsidianFlavoredMarkdown(...),
>   Plugin.GitHubFlavoredMarkdown(),
>   Plugin.TableOfContents(),
>   Plugin.CrawlLinks(...),
>   Plugin.Description(),
> ]
> 
> filters: [
>   Plugin.RemoveDrafts()
> ]
> 
> emitters: [
>   Plugin.AliasRedirects(),
>   Plugin.ComponentResources(...),
>   Plugin.ContentPage(),
>   Plugin.FolderPage(),
>   Plugin.TagPage(),
>   Plugin.ContentIndex(...),
>   Plugin.Assets(),
>   Plugin.Static(),
>   Plugin.NotFoundPage(),
> ]
> ```

Reviewer: ```suggestion
```

Author: Done.

> There are a few more that you can add to the `transformers`, `filters`, and `emitters`:

Reviewer: ```suggestion
There are additional [[transformers]], [[emitters]], that you can use:
```

Options:

- `delimiters`: the delimiters to use for the frontmatter. Can have one value (e.g. `"---"`) or separate values for opening and closing delimiters (e.g. `["---", "~~~"]`). Defaults to `"---"`.
- `language`: the language to use for parsing the frontmatter. Can be `yaml` (default) or `toml`.

### Plugin.CreatedModifiedDate

This plugin determines the created, modified, and published dates for a file using three potential data sources: front matter metadata, Git history, and the filesystem.

If it encounters invalid date formats, it defaults to the current date.

Options:

- `priority`: the data sources to consult for date information. Highest priority first. Possible values are `"frontmatter"`, `"git"`, and `"filesystem"`. Defaults to `"frontmatter", "git", "filesystem"]`.

### Plugin.Latex

This plugin adds LaTeX rendering support. See [[Latex]] for more details.

Options:

- `renderEngine`: the engine to use to render LaTeX equations. Can be `"katex"` for KaTeX or `"mathjax"` for MathJax. Defaults to KaTeX.

### Plugin.SyntaxHighlighting

This plugin adds syntax highlighting for code blocks. See [[syntax highlighting]] for more information.

#### Options

> - `theme`: a separate id of one of the [themes bundled with Shikiji](https://shikiji.netlify.app/themes). One for light mode and one for dark mode. Defaults to `theme: { light: "github-light", dark: "github-dark" }`.

Reviewer: I would prefer this to be a table that links to each plugins docs that we wrote, other having a centralized title here.

So that we don't have to repeat ourself.
